core/chains/evm/log/registrations.go

```diff
@@ -108,14 +140,36 @@ func (r *registrations) removeSubscriber(sub *subscriber) (needsResubscribe bool
 
 	needsResubscribe = handlers.removeSubscriber(sub)
 
-	if len(r.handlersByConfs[sub.opts.MinIncomingConfirmations].handlersByAddr) == 0 {
+	if len(r.handlersByConfs[sub.opts.MinIncomingConfirmations].lookupSubs) == 0 {
 		delete(r.handlersByConfs, sub.opts.MinIncomingConfirmations)
 		r.resetHighestNumConfirmationsValue()
 	}
 
 	return
 }
 
+// checkRemoveSubscriber validates we aren't violating any assumptions by removing this subscriber
+func (r *registrations) checkRemoveSubscriber(sub *subscriber) error {
+	jobID := sub.listener.JobID()
+	if _, exists := r.registeredSubs[sub]; !exists {
+		return errors.Errorf("Cannot remove subscriber %p for job ID %v: not registered", sub, jobID)
+	}
+	delete(r.registeredSubs, sub)
+	addrs, exists := r.jobIDAddrs[jobID]
+	if !exists {
+		return errors.Errorf("Cannot remove subscriber %p: jobIDAddrs was missing job ID %v", sub, jobID)
```

Comment:
If we hit this error here, doesn't it mean something is wrong with our internal data structures?

We were able to find the sub in r.registeredSubs[sub], in above step.
But now we don't find the address for the jobId.

It suggests some data corruption inside the internal structures.

Or is that your intention, to return Panic in such cases, to indicate there's a bug in our registrations internal storage?

Reply:
I originally had it Panic here, but @jmank88 suggested to clean it up by returning error and panicking in one level above

Reply:
I thought it would be cleaner and more easily testable this way, but if the indirection is confusing then could switch back :shrug: - or maybe some kind of `invariant violation:` prefix on these errors could help?

Reply:
If moving the panics back inside, maybe renaming to a `must*` prefix would be fitting?